### PR TITLE
Full Cilogon integration

### PR DESCRIPTION
**Pull request by GeorgianaElena**
This PR adds another way to use CILogon auth for our hubs, one that doesn't require relying on Auth0.

It adds a script, `cilogon_app.py` that: 
* uses the administrative client credentials that we received from CILogon to  `create/udpate/delete/get` approved OAuth applications for our hubs.
* stores and ecrypts the credentials of each hub in `config/clusters/<cluster_name>/enc-<hub_name>.secret.values.yaml

The process to activate CILogon for a hub using the credentials generated above is very similar to the current process of using GitHubOAuthenticator.

Also, this PR enables CILogon for our `2i2c dask-staging` hub and `2i2c demo` hub. It also adds credentials for the 2i2c staging hub also, but without activating it for this hub.

The  [`dask-staging`](https://dask-staging.pilot.2i2c.cloud) hub uses this, so you can play with it. It works from what I've tested so far.

Reference https://github.com/2i2c-org/infrastructure/issues/967

### CODE REVIEW

File: config/clusters/2i2c/enc-cilogon-clients.secret.yaml
Comment (consideRatio) [inline config/clusters/2i2c/enc-cilogon-clients.secret.yaml:1]: This seems confusing to me. Should this be named like that?

File: config/clusters/2i2c/cluster.yaml
Comment (choldgraf) [inline config/clusters/2i2c/cluster.yaml:35]: I like that we are allowing them to co-exist for a bit! As long as we don't have too much headaches maintaining both, this should give us time to try out CILogon without disrupting our current workflows as much

File: deployer/cilogon_auth.py
Comment (choldgraf) [inline deployer/cilogon_auth.py:1]: since we now have several auth-related python modules, maybe we should nest these all under `deployer/auth/`?

**Review by choldgraf — COMMENTED** [review]
This looks nice to me! A few comments and questions. I'll let others on the @2i2c-org/tech-team give a more technical review!

- Could we add the start of some lightweight docs for how we'd use CILogon auth? I think it could be helpful exercise to think about the changes from a user's perspective. It doesn't need to be totally complete since I know you'd like to get the PR in quickly, but something to help people start would be useful.
- Given what you've seen, are there any major decision-points that you can think of where we'd want to use Auth0 vs. CILogon?
- Have you encountered any major downsides or concerns about this approach? Or have you encountered any major benefits?

File: config/clusters/2i2c/enc-cilogon-clients.secret.yaml
Comment (GeorgianaElena) [reply config/clusters/2i2c/enc-cilogon-clients.secret.yaml:1]: This name comes from https://github.com/2i2c-org/infrastructure/blob/6bf2e153256fd15b2793ea8229e1da2b75efe7cc/deployer/hub.py#L135.

We had this convention before too https://github.com/2i2c-org/infrastructure/blob/dd1517d2414f6f49ba22d325074bf488908dbb0e/deployer/hub.py#L131-L133

This is how the 2i2c Auth0 applications are called https://manage.auth0.com/dashboard/us/2i2c/applications.

This client name gets written to this encrypted file [here](https://github.com/2i2c-org/infrastructure/blob/6bf2e153256fd15b2793ea8229e1da2b75efe7cc/deployer/cilogon_auth.py#L160-L170). The reason we are storing these for cilogon and we didn't for Auth0 is because cilogon retrieves details about a client only if provided with the client's id and not its name (`2i2c-dask-staging` is such a name). Also, it doesn't retrieve you the client's secret afterwards, only when it gets created.

File: deployer/cilogon_auth.py
Comment (sgibson91) [reply deployer/cilogon_auth.py:1]: Yes, I would be +1 on this too. _But_ happy for that to be done in a separate PR so as to not slow this one down, so long as we track it.

Comment (GeorgianaElena) [reply deployer/cilogon_auth.py:1]: I've added two lists in the top level comments to track what I believe should be happening as part of this PR and what should be their own separate issues after we merge this. WDYT?

Comment (choldgraf) [reply deployer/cilogon_auth.py:1]: Looks good to me! For the docs, I think the minimal set would be the steps a team member would need to take to:

- Set up CILogon auth with a hub (that used this approach instead of `auth0`)
- Any common next steps in setting up auth (e.g. white-list one or more organizations, or specific users...basically turning our big "regexes" into more structured config)

**Review by yuvipanda — CHANGES_REQUESTED** [review]
Thank you so much for working on this, @GeorgianaElena!!!!

The primary reason the auth0 stuff was baked into the deployer was so we won't have to keep secret files in the repo. However, given that is not possible with CILogon (we can not retrieve secret key multiple times), I think we should instead change this to something where the script just spits out appropriate sops encrypted files as `enc-<hub-name>.secret.values.yaml` file (just using the CILogon authenticator), which we can just check-in. This also simplifies the deployer (once we can get rid of auth0 at least), and enhances right to replicate.

So we could have something like a new-hub script or cookie-cutter that talks to the CILogon API, generates the appropriate secrets and puts it in place. I think the code for this already exists here.

Once we move off auth0 completely, we can just get rid of the dynamic auth part of the deployer completely.

**Comment by GeorgianaElena** [comment]
Thanks for the feedback @yuvipanda. I just pushed some changes with the cilogon client provider turned into a script.
It's not ready, but I would appreciate an early feedback so that I don't get into another rabbit hole.

Also, weird thing, I just got a 500 error with a message saying that the cilogon credentials I was trying to use don't have an "admin" role. @choldgraf do you mind asking the CILogon folks if they've revoked our admin credentials? I'm afraid they think we've abused their system since I've created quite a few dummy clients while testing the implementation 😕

**Comment by choldgraf** [comment]
@GeorgianaElena will send them an email today!

**Comment by GeorgianaElena** [comment]
Thanks a lot @choldgraf :blossom:

**Comment by yuvipanda** [comment]
@GeorgianaElena i haven't done a deep dive into the code but overall this seems the right direction to me!

**Comment by GeorgianaElena** [comment]
I have marked this as ready for review again!

I still want to add some things like:
- [ ] ability to delete a client/secret pair (i.e. OAuth application)
- [ ] docs about the `cilogon_app.py` script
- [ ] update docs on how to update/delete a hub that's using CILogon using this script

But I don't think it should block a first review iteration.

Also, note that the pre-commit is also failing on the main branch, so it shouldn't be because of the changes here.

File: docs/howto/configure/auth-management.md
Comment (yuvipanda) [inline docs/howto/configure/auth-management.md:280]: Is there an issue for this upstream? If not, let's file one and link to it here?

**Review by yuvipanda — CHANGES_REQUESTED** [review]
This is pretty awesome! I left some minor style and follow-up comments, but I think this is ready to merge as is once those are fixed!

File: config/clusters/2i2c/enc-cilogon-clients.secret.yaml
Comment (consideRatio) [reply config/clusters/2i2c/enc-cilogon-clients.secret.yaml]: Ah thanks for the thorough comment and references @GeorgianaElena! Sorry for the slow followup on this! I see now that the structure has changes, so I figure this may be a resolved comment.

---

I think a key confusion point for me was regarding that this was a .YAML file that we load, but it wasn't a helm chart values file and didn't seem to be a cluster.yaml like file either - so then it seemed like there wasn't a schema to validate this configuration.

I think that for any configuration it is probably good if we avoid having keys that contain data by themselves though, so instead of having a name as a key like...

```yaml
this-is-some-cluster-hub-name: some value
```

We could have something like....

```yaml
clusterHub:
  name: this-is-some-cluster-hub-name
  someProperty: some-value
```

This latter would be easy to have a schema for, and its data would be described in the YAML structure etc.

Comment (consideRatio) [reply config/clusters/2i2c/enc-cilogon-clients.secret.yaml]: I see https://github.com/2i2c-org/infrastructure/pull/1089#pullrequestreview-912171302 which is related, and yes - this seems to be resolved!

File: docs/howto/configure/auth-management.md
Comment (GeorgianaElena) [reply docs/howto/configure/auth-management.md]: Fixed in https://github.com/2i2c-org/infrastructure/pull/1089/commits/6819a9e6a13fb9e3bbdf7fdda21809ad86cf9c68

File: docs/howto/configure/auth-management.md
Comment (GeorgianaElena) [reply docs/howto/configure/auth-management.md:296]: Done in https://github.com/2i2c-org/infrastructure/pull/1089/commits/249e8a62aed3eadd5e97f569f750fae3848a26f9

File: docs/howto/configure/auth-management.md
Comment (GeorgianaElena) [reply docs/howto/configure/auth-management.md]: Done in https://github.com/2i2c-org/infrastructure/pull/1089/commits/d607b333a09c0fd61fd5bf535fad1dc9fcbdf303

File: deployer/cilogon_app.py
Comment (GeorgianaElena) [reply deployer/cilogon_app.py]: Thanks @yuvipanda! I agree with the points above and fixed them in https://github.com/2i2c-org/infrastructure/pull/1089/commits/2c49f3cd4d81114b9b1ce47f2b6096f0c43633e3

**Comment by GeorgianaElena** [comment]
This should be now ready for a new review 👀 or a merge

File: docs/howto/configure/auth-management.md
Comment (yuvipanda) [inline docs/howto/configure/auth-management.md:303]:
```suggestion
  - Check the [CILogon scopes section](https://www.cilogon.org/oidc#h.p_PEQXL8QUjsQm) to checkout available values for `USERNAME_KEY` claim. 
     Be careful selecting a value, as once picked this can not be easily changed! It should ideally be something
     that the end user *can not change*, as otherwise they can change it after logging in to access other
     users' servers! 
```

**Comment by yuvipanda** [comment]
@GeorgianaElena i've left some suggested changes in https://github.com/GeorgianaElena/pilot-hubs/pull/2! I'm going to now test deploying this on the demo hub.

**Comment by yuvipanda** [comment]
https://github.com/GeorgianaElena/pilot-hubs/pull/3 is me trying to deploy this to the demo hub, but ran into problems :( I've detailed them in the PR. I think if we can resolve this, this is good to go.

This is such a huge step forward, *thank you* for working on this @GeorgianaElena!

**Comment by yuvipanda** [comment]
The test failure is a link to a file introduced in this PR, so is alright to let fail.

File: docs/howto/configure/auth-management.md
Comment (GeorgianaElena) [reply docs/howto/configure/auth-management.md]: @consideRatio, they were added to the config directly. See https://github.com/GeorgianaElena/pilot-hubs/pull/2#discussion_r834383513

Comment (consideRatio) [reply docs/howto/configure/auth-management.md]: Oh nevermind, i see that there are default scopes provided in the basehub helm chart - okay!

Overall I value duplicating configuration if it helps me understand what is actually configured in a single place. For example, I may look at this configuration and be surprised that it works - should it have some scopes specified? To understand it properly, I'd need to know if scopes are appended by `hub.extraConfig` or via `hub.config` as specified in the basehub helm chart default values etc.

Comment (consideRatio) [reply docs/howto/configure/auth-management.md]: @GeorgianaElena @yuvipanda what do you think about where we put config in a situation like this? Do you value needing to provide less config and avoiding duplicating configuration across files more than you value the complexity of dispersing the config in different places?

Comment (yuvipanda) [reply docs/howto/configure/auth-management.md]: @consideRatio Thank you for raising this question! I definitely value needing to provide less config and avoiding duplicating configuration across files more than you value the complexity of dispersing the config in different places? I think multiple helm chart values files that cascade is a very important part of keeping config concise. I'll write up a doc on 'how to find what config', which might help - as for example, you might sometimes need to look at defaults in z2jh or even in kubespawner if necessary!

Comment (yuvipanda) [reply docs/howto/configure/auth-management.md]: I opened https://github.com/2i2c-org/infrastructure/issues/1144, @consideRatio

Comment (consideRatio) [reply docs/howto/configure/auth-management.md]: Thank you Yuvi, I'll mark this as resolved to not bog down this PR with long term deliberations!

**Review by yuvipanda — APPROVED** [review]
@GeorgianaElena this is super awesome, am very excited to get this to land!

1. https://github.com/GeorgianaElena/pilot-hubs/pull/4 keeps status quo for demo hub, and we can figure out what to do with IDPs
2. https://github.com/GeorgianaElena/pilot-hubs/pull/5 removes some extra config
3. Can you update the PR description here too?

I think you can merge after these two are done.

File: docs/howto/configure/auth-management.md
Comment (yuvipanda) [reply docs/howto/configure/auth-management.md]: Turns out you don't actually need to set scopes at all, as these are the same that's [the default](https://github.com/jupyterhub/oauthenticator/blob/8c1cf0c616d5e19d1f93238673f17db2daf57fa0/oauthenticator/cilogon.py#L69) set by CILogonOAuthenticator! https://github.com/GeorgianaElena/pilot-hubs/pull/5 removes these.

**Comment by GeorgianaElena** [comment]
Merging this now! THANK YOU! 🚀

**Comment by damianavila** [comment]
Lovely piece of work, @GeorgianaElena!!

**Pull request merged**
